File: 0529_updateBoard.py
from collections import deque

class Solution:
    def updateBoard(self, board: List[List[str]], click: List[int]) -> List[List[str]]:
        r, c = click
        if board[r][c] == 'M':
            board[r][c] = 'X'
            return board # if the clicked cell is mine, game over!
        m = len(board)
        n = len(board[0])
        count = 0
        for a, b in (r-1,c-1),(r-1,c),(r-1,c+1),(r,c-1),(r,c+1),(r+1,c-1),(r+1,c),(r+1,c+1):
            if 0 <= a < m and 0 <= b < n and board[a][b] == 'M':
                count += 1
        if count > 0:
            board[r][c] = str(count)
            return board # if there is a mine around the clicked cell, the cell turns to the digit showing count of mine(s) around the cell.
        Q = deque([(r, c)])
        visited = set()
        while Q:
            r, c = Q.popleft()
            board[r][c] = 'B'
            if (r, c) not in visited:
                visited.add((r, c))
                for i, j in (r-1,c-1),(r-1,c),(r-1,c+1),(r,c-1),(r,c+1),(r+1,c-1),(r+1,c),(r+1,c+1):
                    if 0 <= i < m and 0 <= j < n:
                        count = 0
                        for a, b in (i-1,j-1),(i-1,j),(i-1,j+1),(i,j-1),(i,j+1),(i+1,j-1),(i+1,j),(i+1,j+1):
                            if 0 <= a < m and 0 <= b < n and board[a][b] == 'M':
                                count += 1
                        if count == 0:
                            Q.append((i,j))
                        else:
                            board[i][j] = str(count)
        return board # if the clicked cell is blank, all adjacent blank cells and cells with digits showing count of mine(s) around the cell are revealed.

File: 0501-1000/0529_updateBoard.java
class Solution {
    public int[][] directions = {{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}};
    
    public char[][] updateBoard(char[][] board, int[] click) {
        int m = board.length, n = board[0].length;
        int x = click[0], y = click[1];

        if (board[x][y] == 'M') {
            board[x][y] = 'X';
            return board;
        } else {
            dfs(board, m, n, x, y);
            return board;
        }
    }

    public void dfs(char[][] board, int m, int n, int x, int y) {
        int count = 0;
        int row;
        int col;

        for (int[] direction: directions) {
            row = x + direction[0];
            col = y + direction[1];
            if (row >= 0 && row < m && col >= 0 && col < n && board[row][col] == 'M') {
                count++;
            }
        }

        if (count > 0) {
            board[x][y] = (char) (count + '0');
            return;
        } 

        board[x][y] = 'B';

        for (int[] direction: directions) {
            row = x + direction[0];
            col = y + direction[1];
            if (row >= 0 && row < m && col >= 0 && col < n && board[row][col] == 'E') {
                dfs(board, m, n, row, col);
            }
        }
    }
}
